### DISCUSSION TITLE
Question about Klipper updates

### DISCUSSION THREAD

**Discussion by AndyHazz** [1coderookie/Klipper4KobraGoNeo#10] [Q&A] [answered]
After installing this firmware and getting everything working, do you know if it's safe to use the klipper update in mainsail? It gives a clear warning message:

`Control boards running the Klipper firmware may need to be flashed with recompiled firmware before the machine can be put back into service.`

The update showing for me in mainsail is a small one ( v0.11.0-267 > v0.11.0-271) so no real rush to update, but I guess over time that will become more important. Maybe add something to the readme about how to update firmware? This is all fairly new to me and you've made it really easy to get started, just need a bit of info on how to keep it going :)

**Comment by 1coderookie** [comment, marked as answer]
Generally speaking, it always can happen that there are changes made somewhere which somehow break compatibility, that's why that checkbox apperas where you have to click on "I understand the risks" or sth like that before it actually let's you make the update itself.  So if you wanna re-assure yourself as much as possible, you can check the commit history at GH which is linked in that window.  And that's why I didn't and won't mention anything about updates (safe or not) to my infopage or the readme here - ppl should refer to the official Klipper doc then. I'll try to move this issue to the discussions though, so ppl can come across this in the future - so thanks for opening this issue :) 

I personally use to have a look at the commit hostory and then I look out for anything related to "hc32f460" as that's the type of our mainboard being used. But I do it more because I'll see if anything 'interesting' or 'relevant' took place which might be worth mentioning at my infopage in that Klipper section, and not cuz I'm afraid that sth will break compatibility.    

But again, yes, sth like that *might* happen - as with every update of any computer system.  
Just as an example: a few weeks ago or so there was an update of the OS packages shown in Mainsail. Ppl updated and then they couldn't connect to their printer anymore. The reason was sth like a change that took place and the `[mcu] serial: /dev/serial/by-id/`didn't work anymore as that changed to `by-path` or so. Don't really remember exactly, but in the end they got ir working by changing that path in the printer.cfg then.  

So, regarding your question about this specific question - I just updated one of my systems which was even older:  
![Bildschirmfoto von 2023-08-24 14-00-40](https://github.com/1coderookie/Klipper4KobraGoNeo/assets/25869362/6884ce4e-3635-4d53-8ca5-1e22d722bc4f)

Everything still works tho:  
![Bildschirmfoto von 2023-08-24 14-04-50](https://github.com/1coderookie/Klipper4KobraGoNeo/assets/25869362/736242a7-0cec-4d57-bf00-a546ff03ad15)

So *imho* you should be safe making that particular update you were asking about.

Besides that, I strongly suggest to always make backups of your files. Not only because an update *might* cause issues, but cuz of the risk of facing an issue with e.g. the mSD card of your RPi or whatever itself.

**Comment by AndyHazz** [comment]
[firmware.zip](https://github.com/1coderookie/Klipper4KobraGoNeo/files/12440584/firmware.zip)

In case it's any use, I went ahead and updated, and compiled and flashed the new firmware myself as I was interested in the process ... like you say I don't think there's anything significant in v11 for Kobra users (https://www.klipper3d.org/Releases.html) but it seems to work fine for me and perhaps the 'Several bug fixes and code cleanups' are of some small benefit.

![image](https://github.com/1coderookie/Klipper4KobraGoNeo/assets/517731/a6227afd-0c44-40aa-a4bb-e680cf37f0a1)

**Reply by 1coderookie** [reply]
Great, thanks for the feedback Andy!

**Comment by Kellwaxx11** [comment]
Hi all
regarding updating klipper, are you aware of anyone who was able to use the SDCard updates method mentioned on the official guide

[board-definitions](https://www.klipper3d.org/SDCard_Updates.html?#board-definitions)

what is missing the the board config for our MCU, to be more exact the _spi_bus_ and _cs_pin_ values.

@jokubasver ,I hope you can help since you had a custom Marlin firmware built, do you think it is possible to extract those values from Marlin firmware, as I couldn't find any mention of those in any board schematic online.

Thanks in advance.

**Reply by jokubasver** [reply]
Huh, I didn't even know this existed! 

It should be possible, I'll take a look soon, shouldn't be difficult to implement.

**Reply by jokubasver** [reply]
The Trigorilla board uses SDIO for the SD card, so we need to use software SPI and manually reboot the printer after flashing, as described in the docs.

This is the board_def to be added to /klipper/scripts/spi_flash/board_defs.py:

```
    'anycubic-trigorilla-v3.0.6': {
        'mcu': "HC32F460",
        'spi_bus': "swspi",
        'spi_pins': "PC8,PD2,PC12",
        'cs_pin': "PA10",
        'skip_verify': True
    }
```
cs_pin might be PC11. The reason why I'm not sure is because it's not working for me. While flashing, it gives this error:
```
./scripts/flash-sdcard.sh /dev/serial/by-id/usb-1a86_USB_Serial-if00-port0 anycubic-trigorilla-v3.0.6
Flashing /home/jokubas/klipper/out/klipper.bin to /dev/serial/by-id/usb-1a86_USB_Serial-if00-port0
Checking FatFS CFFI Build...
Connecting to MCU...Connected
Checking Current MCU Configuration...Done

SD Card Flash Error: Unknown command: config_spi
Traceback (most recent call last):
  File "/home/jokubas/klipper/scripts/spi_flash/spi_flash.py", line 1647, in main
    spiflash.run()
  File "/home/jokubas/klipper/scripts/spi_flash/spi_flash.py", line 1584, in run
    self.run_reactor_task(self.run_reset_upload)
  File "/home/jokubas/klipper/scripts/spi_flash/spi_flash.py", line 1570, in run_reactor_task
    k_reactor.run()
  File "/home/jokubas/klipper/klippy/reactor.py", line 292, in run
    g_next.switch()
  File "/home/jokubas/klipper/klippy/reactor.py", line 340, in _dispatch_loop
    timeout = self._check_timers(eventtime, busy)
  File "/home/jokubas/klipper/klippy/reactor.py", line 158, in _check_timers
    t.waketime = waketime = t.callback(eventtime)
  File "/home/jokubas/klipper/klippy/reactor.py", line 48, in invoke
    res = self.callback(eventtime)
  File "/home/jokubas/klipper/scripts/spi_flash/spi_flash.py", line 1524, in run_reset_upload
    self.run_sdcard_upload(eventtime)
  File "/home/jokubas/klipper/scripts/spi_flash/spi_flash.py", line 1542, in run_sdcard_upload
    self.mcu_conn.configure_mcu(printfunc=output_line)
  File "/home/jokubas/klipper/scripts/spi_flash/spi_flash.py", line 1359, in configure_mcu
    self._configure_mcu_spibus(printfunc=printfunc)
  File "/home/jokubas/klipper/scripts/spi_flash/spi_flash.py", line 1307, in _configure_mcu_spibus
    self._serial.send(cmd)
  File "/home/jokubas/klipper/klippy/serialhdl.py", line 257, in send
    cmd = self.msgparser.create_command(msg)
  File "/home/jokubas/klipper/klippy/msgproto.py", line 332, in create_command
    self._error("Unknown command: %s", msgname)
  File "/home/jokubas/klipper/klippy/msgproto.py", line 243, in _error
    raise error(self.warn_prefix + (msg % params))
msgproto.error: Unknown command: config_spi
```
So there's something more to it than that. HC32F460 has no SPI includes, so that's probably why. I've tried to mess around with it for a short while, but it doesn't seem simple, so I'm leaving this at that.

**Reply by Kellwaxx11** [reply]
thank you for trying 

i tried your board definition and got the same errors using KIAUH as well.
but what i noticed that the firmware started acting crazy after, connects and disconnects repeatedly for short periods , I had to flash the firmware the old way to get it back to act normal

anyways seems that we need to stick to the old way, thanks again

**Reply by Kellwaxx11** [reply]
do you know of any method to just copy files to the SD card with klipper, so we can manually copy the firmware file and update the old way, except not need to use a card reader (i keep losing that  :) ) ?

**Reply by jokubasver** [reply]
The only way is using the SD card and reader.
